### DISCUSSION TITLE
Why is this .or() query on foreign table not returning foreign table object?

### DISCUSSION THREAD

**Discussion by jimmiejackson414** [supabase/supabase#8116] [Questions] [answered]
I'm working on building a query that allows a user to order, sort and search. Currently, when trying to allow for search, the foreign table is returning `null`.

```javascript
let query = supabase
  .from('friends')
  .select(`*, user:friend_id(*)`, { count: 'exact' })
  eq('user_id', userId);

  // build query filters
  if (search) { query = query.or(`first_name.ilike.%${search}%`, { foreignTable: 'user' }); }
  if (orderBy) { query = query.order(orderBy, sortOptions ?? undefined); }
  if (from && to) { query = query.range(from, to); }

  const { data, error, count } = await query;
  console.log({ data });
  console.log({ error });
  console.log({ count });
```

And this returns:
```
{
  data: [
    {
      id: 1,
      invite_sent_on: null,
      invite_accepted_on: null,
      created_at: '2022-05-25T20:24:39+00:00',
      updated_at: '2022-05-25T20:24:39+00:00',
      user_id: '159147fc-f2b3-4872-8dc5-fc7ec749e55d',
      friend_id: '8d1efb59-3c54-40ca-b378-0428ad7b6029',
      user: null // <-- the issue
    }
  ]
}
{ error: null }
{ count: 1 }
```

If I comment out the line that attaches `.or()` to the query, then the User object from the foreign table is correctly returned. However, when I have it commented back in, you can see that the User object returns null.

What might be causing this?

**Comment by GaryAustin1** [comment, marked as answer]
The .or statement is standalone.  It normally expects two or more filters to do the .or on and returns the result of them all or'd together.  If a foreign table is involved, all expressions are on that table.  The results of the .or are anded with the rest of your filters.

What you are showing is basically anding your foreign table filter with all your other filters as the .or only has one test in it.

I guess I would expect though the whole query to be affected not just the user object returned.

**Reply by jimmiejackson414** [reply]
Eventually I wanted to add a couple more clauses within the .or, I only had one in there temporarily to test it out. Here's what I'd ultimately like to achieve:

```javascript
query = query.or(`first_name.ilike.%${search}%,last_name.ilike.%${search}%,email.ilike.%${search}%`, { foreignTable: 'user' });
```

I'm not sure if it helps at all, but when I console log the query being built, this is the URL that is being generated:
`'https://xxxxxxxxxxxxxxxx.supabase.co/rest/v1/friends?select=*%2Cuser%3Afriend_id%28*%29&user_id=eq.159147fc-f2b3-4872-8dc5-fc7ec749e55d&user.or=%28first_name.ilike.%25jimmie%25%2Clast_name.ilike.%25jimmie%25%2Cemail.ilike.%25jimmie%25%29&user.order=email.asc.nullslast'`

**Reply by GaryAustin1** [reply]
gotcha...
Yeah, at the moment I don't see why either.

**Reply by GaryAustin1** [reply]
Could be this is messing up the URL.
```
ILIKE operator (use * in place of %)
```
From PostgREST.org....

**Reply by jimmiejackson414** [reply]
@GaryAustin1 interesting. I tried replacing the % with *, same result. I also looked at the [Supabase docs](https://supabase.com/docs/reference/javascript/or#use-or-on-foreign-tables) for .or(), and the query isn't surrounded by any special characters so I tried that as well (with the same result).

**Reply by GaryAustin1** [reply]
Everything inside the () of the .or needs to be in PostgREST syntax, so if not this issue, likely would have been an issue.

**Reply by jimmiejackson414** [reply]
I'm not sure though, when I look at the `.or` syntax specifically [in the docs](https://postgrest.org/en/stable/api.html?highlight=embedded#embedded-filters), there's no mention of surrounding the query with any special characters.

**Reply by GaryAustin1** [reply]
From same docs... It is not surrounding with special characters, just use * instead of % in your ilike pattern.
![image](https://user-images.githubusercontent.com/54564956/183157694-9bbfe9cf-cab7-4ed2-837c-23c22b667a9b.png)

Edit:  Just for reference your query cleaned up and  the overall format looks intact for the URL itself:
```
https://xxxxxxxxxxxxxxxx.supabase.co/rest/v1/friends?select=*,user:friend_id(*)&user_id=eq.159147fc-f2b3-4872-8dc5-fc7ec749e55d&user.or=(first_name.ilike.%jimmie%,last_name.ilike.%jimmie%,email.ilike.%jimmie%)&user.order=email.asc.nullslast
```
I do notice you are also ordering on the same foreign table, if I read that correctly.

**Reply by jimmiejackson414** [reply]
Wouldn't that be the case if I were using the `.ilike()` directly, rather than `.or()`? [Supabase Docs](https://supabase.com/docs/reference/javascript/ilike)

**Reply by GaryAustin1** [reply]
Supabase processes the direct ilike I'm pretty sure.
This is the note on .or
https://supabase.com/docs/reference/javascript/or#notes
postgrest-js does not process what is inside the .or()

**Reply by jimmiejackson414** [reply]
Hoo boy, now I'm really confused haha

**Reply by GaryAustin1** [reply]
Yeah I am now too... I don't see any code to process the .ilike pattern in postgrest-js.  
But we have drifted from your immediate problem.

**Reply by GaryAustin1** [reply]
@steve-chavez can you skim thru the op's original problem and clarify the ilike * % versus the .or with ilike * % 
I just ran 
`.ilike('col','*cat*')` and `.ilike('col','%cat%')` and both work.  Did not try .or with ilike in it....

Edit: .or works with both % and * also....

**Reply by steve-chavez** [reply]
@GaryAustin1 Sure. TLDR both `*` and `%` should work the same through `postgrest-js`. 

`*` is mentioned on the PostgREST docs because `%` is a reserved character in URLs and requires [URL encoding](https://en.wikipedia.org/wiki/Percent-encoding)(`%cat%` would have to be expressed as `%25cat%25`). This doesn't matter in some http clients/libraries as they url encode their parameters by default. `postgrest-js` uses [searchParams](https://github.com/supabase/postgrest-js/blob/master/src/lib/PostgrestFilterBuilder.ts#L169-L172) which [already urlencodes](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams).

**Edit**: Clarified this on https://github.com/PostgREST/postgrest-docs/commit/eb100cc7460eef07b6b44ae83b678483e9074089.

**Reply by jimmiejackson414** [reply]
@steve-chavez any idea why the foreign table object isn't being returned in this scenario?

**Reply by steve-chavez** [reply]
@jimmiejackson414 Yeah, you're basically filtering the `user` rows with the `first_name.ilike.%${search}%` condition.

Do you want to filter `friends` based on `user` conditions? If so you could use [!inner](https://supabase.com/docs/reference/javascript/select#filtering-with-inner-joins).

**Reply by jimmiejackson414** [reply]
@steve-chavez brilliant, that might be the trick. Thanks a ton!

**Reply by jimmiejackson414** [reply]
@steve-chavez apologies for reviving this, but I'm noticing that when the User record from the foreign table isn't found, it's still returning the parent Friend record. Is there a way to return null for the Friend record when the User record isn't found? Or will that require a custom stored function?

**Reply by steve-chavez** [reply]
>  but I'm noticing that when the User record from the foreign table isn't found, it's still returning the parent Friend record

Can you provide an example response? That shouldn't happen with `!inner`.

**Reply by jimmiejackson414** [reply]
Sure thing. I'm not able to specifically use the `!inner` keyword so I may be misunderstanding something here, but I'm getting the error about multiple relationships being found, so here is the query I have currently:

```javascript
const { page, size, orderBy, sortOptions, search } = input;
const { from, to } = getPagination(page as number, size as number);

let query = supabase
  .from('friends')
  .select('*, user:users!friends_friend_id_fkey(*)', { count: 'exact' })
  .eq('user_id', userId);

// build query filters
if (search) { query = query.or(`first_name.ilike.${search},last_name.ilike.${search},email.ilike.${search}`, { foreignTable: 'users' }); }
if (orderBy) { query = query.order(orderBy, sortOptions || undefined); }
if (from && to) { query = query.range(from, to); }
// TODO: Currently the .or() method still returns Friend record even if User not found.

const { data, error, count } = await query;
console.log({ data });
console.log({ error });
console.log({ count });
```

And the response I'm getting when searching for a user that shouldn't return a record:
```json
{
  "data": {
    "friends": {
      "friends": [
        {
          "id": "1",
          "user_id": "159147fc-f2b3-4872-8dc5-fc7ec749e55d",
          "friend_id": "8d1efb59-3c54-40ca-b378-0428ad7b6029",
          "invite_sent_on": null,
          "invite_accepted_on": null,
          "created_at": "2022-05-25T20:24:39+00:00",
          "updated_at": "2022-05-25T20:24:39+00:00",
          "user": null // <-- foreign record is null, shouldn't return parent record
        }
      ],
      "count": 1,
      "errors": null
    }
  }
}
```

**Reply by steve-chavez** [reply]
>  // <-- foreign record is null, shouldn't return parent record
> I'm not able to specifically use the !inner keyword 

Hm, the above wouldn't happen with `!inner`. Why are you not able to use it? Should be possible like:

```js
let query = supabase
  .from('friends')
  .select('*, user:users!friends_friend_id_fkey!inner(*)', { count: 'exact' })
  .eq('user_id', userId);
```

**Reply by jimmiejackson414** [reply]
@steve-chavez ah, so it was a misunderstanding on my part! I thought that the `friends_friend_id_fkey` was a more specific naming convention for `inner`. What you have above works beautifully, I really appreciate your help!

**Reply by steve-chavez** [reply]
Cool! The `!` thing represent options for embedding another table, these now use the raw REST syntax but there have been some discussions to abstract that in the js library.
